### PR TITLE
Extend warnings

### PR DESCRIPTION
**Pull request by chriseppstein**
In porting from base classes to placeholder selectors, I really needed this.

In 3.2 we will warn if a selector being extended is not found. In 3.3/4.0 it will become an error.

The warning can be silenced by placing `!optional` at the end of an `@extend statement`. This maybe useful for libraries and partial re-use.

### CODE REVIEW

File: lib/sass/engine.rb [hunk 1]
```diff
@@ -671,7 +671,12 @@ def parse_directive(parent, line, root)
         raise SyntaxError.new("Illegal nesting: Nothing may be nested beneath extend directives.",
           :line => @line + 1) unless line.children.empty?
         offset = line.offset + line.text.index(value).to_i
-        Tree::ExtendNode.new(parse_interp(value, offset))
+        optional = false
+        if value =~ /\!optional\s*$/
+          optional = true
+          value = value.gsub(/\s*!optional\s*$/,'')
+        end
```
Comment (chriseppstein) [inline lib/sass/engine.rb:683]: Should this use the `RX::OPTIONAL` constant?

File: lib/sass/selector/abstract_sequence.rb [hunk 2]
```diff
@@ -64,6 +64,43 @@ def has_placeholder?
...
+
+      # Marks this selector as not requiring the base selector to exist.
+      def extension_optional!
+        @extension_optional = true
+      end
+
+      # Checks whether this selector has been extended
+      def extension_optional?
+        @extension_optional
+      end
```
Comment (chriseppstein) [inline lib/sass/selector/abstract_sequence.rb:93]: I thought about refactoring these four methods into a shared module for `Selector::AbstractSequence` and `Selector::Simple`. Wasn't sure where to put that or what to call it. Alternatively there could be a shared base class for the two kinds of selectors.

File: lib/sass/selector/sequence.rb [hunk 3]
```diff
@@ -20,6 +21,7 @@ def line=(line)
       # @param filename [String, nil]
       # @return [String, nil]
       def filename=(filename)
+        @filename = filename
```
Comment (chriseppstein) [inline lib/sass/selector/sequence.rb:24]: this bug baffled me for quite a while. >_<

File: lib/sass/tree/root_node.rb [hunk 4]
```diff
@@ -20,8 +20,24 @@ def render
...
         result, extends = Visitors::Cssize.visit(result)
-        result = result.do_extend(extends) unless extends.empty?
-        result.to_s
+        result.do_extend(extends).to_s
+      end
+
+      def do_extend(extends)
+        return self if extends.empty?
+        result = super
+        extends.each do |e, by|
```
Comment (chriseppstein) [inline lib/sass/tree/root_node.rb]: It seems like the way that this data structure works, it's not possible for this part of the code to properly handle a missing `@extend .foo.bar` because `.foo` and `.bar` are stored as separate keys in the map. As a result, the user gets two warnings in this case. There might be a way to change how the selectors are stored or how each works that would allow for the correct error message in this case. But since this is not a common case, I'm willing to ship with this issue in place. Having slightly confusing warnings in this case is still better than not knowing.

File: lib/sass/util/subset_map.rb [hunk 5]
```diff
@@ -96,6 +95,14 @@ def get(set)
       def [](set)
         get(set).map {|v, _| v}
       end
+
+      def each
+        @hash.each do |k,v|
+          v.each do |(_,_,i)|
```
Comment (chriseppstein) [inline lib/sass/util/subset_map.rb]: do you think it would it be correct if I only yielded once here with v being a map of indexes to values.

**Comment by chriseppstein** [comment]
I have dusted this branch off and am very pleased with it now.

It only gives a single warning for a single unused extend even when multiple selectors are being extended.

You had given a case that you were concerned about (that comment was lost when I rebased against master) but it amounted to:

``` scss
.foo { a: b; }
.bar { c: d; }
.baz { @extend .foo.bar; }
```

Which you claimed wouldn't warn, but it does -- the extend didn't apply and the warn occurs like so:

```
WARNING on line 3:
  Missing selector .foo.bar not found for @extend of .baz.
  This will become an error in a future release.
  To allow this condition, change to:
    @extend .foo.bar !optional
```

I'd like to merge this in.

**Comment by chriseppstein** [comment]
Rebased on master so that it applies cleanly.

File: lib/sass/engine.rb (hunk 1 above)
Comment (nex3) [inline lib/sass/engine.rb:683]: I don't think it needs to, although `optional = value.gsub!(/\s*!optional\s*$/, '')` would be cleaner.

File: lib/sass/scss/parser.rb [hunk 6]
```diff
@@ -289,7 +289,7 @@ def else_directive
       end
 
       def extend_directive
-        node(Sass::Tree::ExtendNode.new(expr!(:selector_sequence)))
+        node(Sass::Tree::ExtendNode.new(expr!(:selector_sequence), !!tok(OPTIONAL)))
```
Comment (nex3) [inline lib/sass/scss/parser.rb:292]: This should consume any whitespace after `!optional`.

File: lib/sass/selector/abstract_sequence.rb [hunk 7]
```diff
@@ -82,6 +82,26 @@ def specificity
         _specificity(members)
       end
 
+      # Marks this selector as not requiring the base selector to exist.
+      def extension_optional!
+        @extension_optional = true
+      end
+
+      # Checks whether this selector has been extended
```
Comment (nex3) [inline lib/sass/selector/abstract_sequence.rb:90]: Wrong docstring

File: lib/sass/selector/sequence.rb (hunk 3 above)
Comment (nex3) [inline lib/sass/selector/sequence.rb:24]: Should this change be separated into its own commit?

**Comment by nex3** [comment]
This patch still has false positives. Example:

``` scss
.foo {@extend .bar.baz}
.bar {a: b}
.baz {c: d}
```

This shouldn't trigger a warning, because both `.bar` and `.baz` exist in the document. I think this was adequately established in our last discussion about this behavior.

This is a fundamental issue with the approach in this patch. Effectively what you're doing is marking each individual `@extend` when it matches a selector, but that's never going to account for valid extendees that just empirically match nothing.

I think we have to fall back on a placeholder approach.
